hikari/auto_mod.py

```diff
@@ -0,0 +1,286 @@
+from hikari.internal import enums
+
+if typing.TYPE_CHECKING:
+    import datetime
+
+    from hikari import traits
+
+
+class AutoModActionType(int, enums.Enum):
+    """The type of an auto-moderation rule action."""
+
+    BLOCK_MESSAGE = 1
+    """Block the content of the triggering message."""
+
+    SEND_ALERT_MESSAGE = 2
+    """Log the triggering user content to a specified channel."""
+
+    TIMEOUT = 3
+    """Timeout the triggering message's author for a specified duration.
+
+    This type can only be set for `KEYWORD` and `MENTION_SPAM` rules.
+
+    Requires the `MODERATE_MEMBERS` permission to use.
+    """
+
+    BLOCK_MEMBER_INTERACTION = 4
+    """Prevents a member from using text, voice, or other interactions."""
+
+
+@attrs_extensions.with_copy
+@attrs.define(kw_only=True, weakref_slot=False)
+class PartialAutoModAction:
+    """Base class for an action which is executed when a rule is triggered."""
+
+    type: AutoModActionType = attrs.field()
+    """The type of auto-moderation action."""
+
+
+@attrs.define(kw_only=True, weakref_slot=False)
+class AutoModBlockMessage(PartialAutoModAction):
+    """Block the content of the triggering message."""
+
+
+@attrs.define(kw_only=True, weakref_slot=False)
+class AutoModSendAlertMessage(PartialAutoModAction):
+    """Log the triggering content to a specific channel."""
+
+    channel_id: snowflakes.Snowflake = attrs.field()
+    """ID of the channel to log trigger events to."""
+
+
+@attrs.define(kw_only=True, weakref_slot=False)
+class AutoModTimeout(PartialAutoModAction):
+    """Timeout the triggering message's author for a specified duration.
+
+    This type can only be set for `KEYWORD` and `MENTION_SPAM` rules.
+
+    Requires the `MODERATE_MEMBERS` permission to use.
+    """
+
+    duration: datetime.timedelta = attrs.field()
+    """The total seconds to timeout the user for (max 2419200 seconds/4 weeks)."""
+
+
+@attrs.define(kw_only=True, weakref_slot=False)
+class AutoModBlockMemberAction(PartialAutoModAction):
+    """Prevents a member from using text, voice, or other interactions."""
+
+
+class AutoModEventType(int, enums.Enum):
+    """Type of event to check for an auto-moderation rule."""
+
+    MESSAGE_SEND = 1
+    """When a member sends or edits a message in the guild."""
+
+    MEMBER_UPDATE = 2
+    """When a member updates their guild or user profile."""
+
+
+class AutoModTriggerType(int, enums.Enum):
+    """Type of trigger for an auto-moderation rule."""
+
+    KEYWORD = 1
+    """Match message content against a list of keywords and regexs."""
+
+    SPAM = 3
+    """Discord's guild anti-spam system."""
+
+    KEYWORD_PRESET = 4
+    """Discord's preset keyword triggers."""
+
+    MENTION_SPAM = 5
+    """Match messages that exceed the allowed limit of role or user mentions."""
+
+    MEMBER_PROFILE = 6
+    """Match user profiles against a list of keywords and regexs."""
+
+
+class AutoModKeywordPresetType(int, enums.Enum):
```

Comment:
Could we reorganize this file a bit and have all the enums at the top and the classes bellow?